plugins/tableselection/plugin.js

```diff
@@ -889,6 +858,56 @@
 		setTimeout( function() {
 			editor.fire( 'afterPaste' );
 		}, 0 );
+
+		function isCustomPaste( selection, selectedCells, pastedTable, boundarySelection ) {
+			var ranges = selection.getRanges(),
+				table = ranges.length && ranges[ 0 ]._getTableElement( { table: 1 } );
+
+			// Do not customize paste process in following cases:
+			// 1. No cells are selected.
+			if ( !selectedCells.length ) {
+				return false;
+			}
+
+			// 2. Table is ignoring tableselection (#2945).
+			if ( table && table.hasAttribute( ignoredTableAttribute ) ) {
+				return false;
+			}
+
+			// 3. It's a boundary position but with no table pasted.
+			if ( boundarySelection && !pastedTable ) {
+				return false;
+			}
+
+			// 3.1 If it's boundary position and table is pasted, proceed with custom paste.
+			if ( boundarySelection ) {
```

Comment:
This looks a little confusing (all other conditions returns `false`) - I see it's intended (https://github.com/ckeditor/ckeditor4/pull/3561#issuecomment-557612461), however some more explicit comment why it needs such early return here could be very useful.

Apart from the explanation, I wonder if you couldn't merge it into last `if` like:

```js
if ( !boundarySelection && !rangeContainsTableElement( ranges[ 0 ] ) ) { ... }
```